examples/hubert/dataset/dataset.py

```diff
@@ -0,0 +1,214 @@
+        phase_type: str,
+        min_sample: int,
+        max_sample: int,
+    ) -> Tuple[List[Path], List[int], List[int]]:
+        """Get the list of paths for iteration.
+        Args:
+            tsv_dir (Path): The root directory of the ``.tsv`` file list.
+            dataset (str): The dataset for training. Options: [``librispeech``, ``librilight``].
+            phase_type (str): The type of the dataset. Options: [``train``, ``valid``].
+            min_sample (int): The minimum number of audio samples in the dataset.
+            max_sample (int): The maximum number of audio samples in the dataset.
+
+        Returns:
+            (list) List of file paths.
+            (list) List of indices that qualify ``min_sample`` <= length <= ``max_sample``.
+            (list) List of waveform lengths.
+        """
+        f_ind_len_list = []
+        with open(tsv_dir / f"{dataset}_{phase_type}.tsv") as f:
+            root = f.readline().rstrip()
+            for index, line in enumerate(f):
+                path, nsample = line.split("\t")
+                path = f"{root}/{path}"
+                nsample = int(nsample)
+                if min_sample <= nsample <= max_sample:
+                    f_ind_len_list.append((path, index, nsample))
+        f_ind_len_list.sort(key=lambda x: x[2])  # sort the file lists by the sequence length
+        f_list, ind_list, len_list = [], [], []
+        for ele in f_ind_len_list:
+            f_list.append(ele[0])
+            ind_list.append(ele[1])
+            len_list.append(ele[2])
+        return f_list, ind_list, len_list
+
+    def _load_audio(
+        self,
+        index: int
+    ) -> Tensor:
+        wav_path = self.f_list[index]
+        waveform, sample_rate = torchaudio.load(wav_path)
+        assert waveform.shape[1] == self.len_list[index]
+        return waveform
+
+    def _load_label(
+        self,
+        label_dir: Path,
+        dataset: str,
+        phase_type: str
+    ) -> List[List[int]]:
+        with open(label_dir / f"{dataset}_{phase_type}.pt") as f:
+            labels = [line.rstrip() for line in f]
+            labels = [[int(ele) for ele in labels[i].split()] for i in self.ind_list]
+        return labels
+
+    def __getitem__(self, index):
+        waveform = self._load_audio(index)
+        length = waveform.shape[1]
+        label = torch.Tensor(self.labels[index])
+        return (waveform, length, label)
+
+
+def collate_fn_hubert(
+    batch
+) -> Tuple[Tensor, Tensor, Tensor]:
+    waveforms, lengths, labels = [], [], []
+    for sample in batch:
+        waveform, length, label = sample
+        waveforms.append(waveform)
+        lengths.append(length)
+        labels.append(label)
+
+    data = torch.zeros(len(batch), max(lengths))
+    for i in range(len(waveforms)):
+        data[i][0:waveforms[i].shape[1]] = waveforms[i][0]
+    lengths = torch.Tensor(lengths)
+    labels = torch.nn.utils.rnn.pad_sequence(labels, batch_first=True)
```

Comment:
Would this op be stateless? I mean we would not do backpropogation here.

Reply:
Could you elaborate on the stateless? The intention here is to compute the NCE loss batch-wise and apply masking afterwards.